src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState;
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        newState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    currentState = newState;
```

Comment:
Instead of assigning `newState` to `currentState` after each action, you could directly modify `currentState` within the switch cases and then push it to the `stateHistory` array. This would follow the DRY principle more closely.